src/QMCWaveFunctions/tests/test_DiracMatrixComputeCUDA.cpp

```diff
@@ -0,0 +1,201 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Laboratory
+//
+// File created by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Laboratory
+//////////////////////////////////////////////////////////////////////////////////////
+
+#include <catch.hpp>
+#include <algorithm>
+#include "Configuration.h"
+#include "OhmmsData/Libxml2Doc.h"
+#include "OhmmsPETE/OhmmsMatrix.h"
+#include "OhmmsPETE/OhmmsVector.h"
+#include "QMCWaveFunctions/WaveFunctionComponent.h"
+#include "QMCWaveFunctions/Fermion/DiracMatrixComputeCUDA.hpp"
+#include "QMCWaveFunctions/tests/makeRngSpdMatrix.hpp"
+#include "Utilities/for_testing/checkMatrix.hpp"
+#include "Utilities/for_testing/RandomForTest.h"
+#include "Platforms/PinnedAllocator.h"
+#include "Platforms/CUDA/CUDALinearAlgebraHandles.h"
+#include "Platforms/tests/test_kernel.hpp"
+
+// Legacy CPU inversion for temporary testing
+#include "QMCWaveFunctions/Fermion/DiracMatrix.h"
+
+
+namespace qmcplusplus
+{
+#ifdef ENABLE_OFFLOAD
+  template<typename T>
+  using OffloadPinnedAllocator = OMPallocator<T, PinnedAlignedAllocator<T>>;
+#elif ENABLE_CUDA
+  template<typename T>
+  using OffloadPinnedAllocator = DualAllocator<T, CUDAAllocator<T>, PinnedAlignedAllocator<T>>;
+#endif
+
+template<typename T>
+using OffloadPinnedMatrix = Matrix<T, OffloadPinnedAllocator<T>>;
+template<typename T>
+using OffloadPinnedVector = Vector<T, OffloadPinnedAllocator<T>>;
+
+
+TEST_CASE("DiracMatrixComputeCUDA_cuBLAS_geam_call", "[wavefunction][fermion]")
+{
+  OffloadPinnedMatrix<double> mat_a;
+  int n = 4;
+  mat_a.resize(n, n);
+  OffloadPinnedMatrix<double> temp_mat;
+  temp_mat.resize(n, n);
+  OffloadPinnedMatrix<double> mat_c;
+  mat_c.resize(n, n);
+
+  DeviceValue<double, OffloadPinnedAllocator<double>> dev_one(1.0);
+  DeviceValue<double, OffloadPinnedAllocator<double>> dev_zero(0.0);
+  
+  std::vector<double> A{2, 5, 8, 7, 5, 2, 2, 8, 7, 5, 6, 6, 5, 4, 4, 8};
+  std::copy_n(A.begin(), 16, mat_a.data());
+  auto cuda_handles = std::make_unique<CUDALinearAlgebraHandles>();
+  int lda= n;
+  cublasSetPointerMode(cuda_handles->h_cublas, CUBLAS_POINTER_MODE_DEVICE);
```

Comment:
Why do we need to test CUBLAS_POINTER_MODE_DEVICE which we don't use?
Is this the reason why DeviceValue is added?

The mode in CUDA is global, we should only use CUBLAS_POINTER_MODE_HOST.

Reply:
It isn't and there should be a handle per thread/crowd.

I disliked the inconsistent semantics for pointers in the cuBLAS API with CUBLAS_POINTER_MODE_HOST.
I don't want to change this until after my benchmarking PR is merged since I have good performance now. I want to be able to track the effect of changes to the device computation, batching, and transfer.

Reply:
I don't think it matters to the performance. I'm wondering if this can help remove the need of DeviceValue.

Reply:
OK. We can make the change after your benchmark PR before connect to the production code path.